### PR TITLE
[ConstantRange][LVI] Add initial support for `multiplyWithNoWrap`

### PR DESCRIPTION
**Pull request by antoniofrighetto**
Introduce support for computing multiplication ranges when nowrap flags are known. This is achieved by intersecting the multiplication range with the saturating one. Note that we may still conservatively return overdefined when handling non-wrapped/non-sign-wrapped ranges.

### CODE REVIEW

File: llvm/lib/IR/ConstantRange.cpp [hunk 1]
```diff
@@ -1167,6 +1169,28 @@ ConstantRange::multiply(const ConstantRange &Other) const {
   return UR.isSizeStrictlySmallerThan(SR) ? UR : SR;
 }
 
+ConstantRange
+ConstantRange::multiplyWithNoWrap(const ConstantRange &Other,
+                                  unsigned NoWrapKind,
+                                  PreferredRangeType RangeType) const {
+  if (isEmptySet() || Other.isEmptySet())
+    return getEmpty();
+
```
Comment (dtcxzyw) [inline llvm/lib/IR/ConstantRange.cpp:1178]: It would be better to early exit when both ranges are full.

File: llvm/lib/IR/ConstantRange.cpp [hunk 2]
```diff
@@ -1167,6 +1169,28 @@ ConstantRange::multiply(const ConstantRange &Other) const {
...
+  ConstantRange Result = multiply(Other);
+
+  // TODO: Bounds are not yet precise if the input sets are non-wrapped /
+  // non-sign-wrapped, as we may still conservatively return full-set in such
+  // cases. To be extended.
+
+  if (NoWrapKind & OverflowingBinaryOperator::NoSignedWrap)
+    Result = Result.intersectWith(smul_sat(Other), RangeType);
```
Comment (dtcxzyw) [inline llvm/lib/IR/ConstantRange.cpp:1186]: @nikic Can you test the compile-time impact?

File: llvm/lib/IR/ConstantRange.cpp [hunk 3]
```diff
@@ -1167,6 +1169,28 @@ ConstantRange::multiply(const ConstantRange &Other) const {
...
+ConstantRange
+ConstantRange::multiplyWithNoWrap(const ConstantRange &Other,
+                                  unsigned NoWrapKind,
+                                  PreferredRangeType RangeType) const {
+  if (isEmptySet() || Other.isEmptySet())
+    return getEmpty();
+
+  ConstantRange Result = multiply(Other);
+
+  // TODO: Bounds are not yet precise if the input sets are non-wrapped /
```
Comment (XChy) [inline llvm/lib/IR/ConstantRange.cpp:1181]: Could you please provide a negative test or an example? It's hard for me to imagine it.

Comment (antoniofrighetto) [reply llvm/lib/IR/ConstantRange.cpp:1181]: For example, given input ranges `[0,2)` and `[2,0)` (bitwidth: 4), we currently return `full-set`, but a more refined CR would be `[2,1)`.

**Comment by antoniofrighetto** [comment]
TBH I have been having a hard time understanding some of the failures that arise when testing exhaustively mul nuw w/ optimality `CheckNonWrappedOnly`. For instance, for input ranges `[0,5)`, `[6,-1)` (bitwidth 4), I get the following:
```
Failure
Value of: NotPreferred(PossibleCR)
  Actual: false (Inputs = [0,5), [6,-1), CR = full-set, BetterCR = [6,1))
Expected: true

Failure
Value of: NotPreferred(PossibleCR)
  Actual: false (Inputs = [0,5), [6,-1), CR = full-set, BetterCR = [0,-1))
Expected: true
```
Isn't the union of the two sets full-set? What am I missing?

**Comment by nikic** [comment]
@antoniofrighetto 0,1,2,3,4 *nuw 6,7,8,9,10,11,12,13,14 = 0,6,7,8,9,10,11,12,13,14

I don't think having accurate mul ranges is feasible without a brute force evaluation. You can drop the TODO.

File: llvm/test/Transforms/SCCP/range-mul-nuw-nsw-flags.ll [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+entry:
+  %cmp = icmp ne i32 %a, 0
+  br i1 %cmp, label %then, label %else
+then:
+  %mul = mul nuw nsw i32 %a, 10000 ; Refined range via mul_nuw: [10000, 0)
+  %add = add nsw i32 %mul, -5000   ; Range: [5000, UINT_MAX - 5000 + 1)
+  %cond = icmp ult i32 %add, 4999
+  ret i1 %cond
+else:
+  call void @opaque()
```
Comment (nikic) [inline llvm/test/Transforms/SCCP/range-mul-nuw-nsw-flags.ll:28]: This call is unnecessary.

File: llvm/unittests/IR/ConstantRangeTest.cpp [hunk 5]
```diff
@@ -1019,6 +1019,99 @@ TEST_F(ConstantRangeTest, Multiply) {
...
+
+  EXPECT_EQ(Empty.multiplyWithNoWrap(Some, OBO::NoUnsignedWrap), Empty);
+  EXPECT_EQ(Some.multiplyWithNoWrap(Empty, OBO::NoUnsignedWrap), Empty);
+  EXPECT_EQ(Full.multiplyWithNoWrap(Full, OBO::NoUnsignedWrap), Full);
+  EXPECT_EQ(Full.multiplyWithNoWrap(Some, OBO::NoUnsignedWrap), Full);
+  EXPECT_EQ(Some.multiplyWithNoWrap(Full, OBO::NoUnsignedWrap), Full);
+  EXPECT_EQ(ConstantRange(APInt(4, 0), APInt(4, 2))
+                .multiplyWithNoWrap(ConstantRange(APInt(4, 2), APInt(4, 0)),
+                                    OBO::NoUnsignedWrap),
+            ConstantRange(4, true));
```
Comment (nikic) [inline llvm/unittests/IR/ConstantRangeTest.cpp:1033]:
```suggestion
            ConstantRange(4));
```
Not signed.

File: llvm/unittests/IR/ConstantRangeTest.cpp [hunk 6]
```diff
@@ -1019,6 +1019,99 @@ TEST_F(ConstantRangeTest, Multiply) {
...
+                                    OBO::NoUnsignedWrap),
+            ConstantRange(8, false));
+
+  EXPECT_EQ(Empty.multiplyWithNoWrap(Some, OBO::NoSignedWrap), Empty);
+  EXPECT_EQ(Some.multiplyWithNoWrap(Empty, OBO::NoSignedWrap), Empty);
+  EXPECT_EQ(Full.multiplyWithNoWrap(Full, OBO::NoSignedWrap), Full);
+  EXPECT_EQ(Full.multiplyWithNoWrap(Some, OBO::NoSignedWrap), Full);
+  EXPECT_EQ(Some.multiplyWithNoWrap(Full, OBO::NoSignedWrap), Full);
+  EXPECT_EQ(ConstantRange(APInt(4, 0), APInt(4, 4))
+                .multiplyWithNoWrap(ConstantRange(APInt(4, -5), APInt(4, 4)),
```
Comment (nikic) [inline llvm/unittests/IR/ConstantRangeTest.cpp:1053]:
```suggestion
                .multiplyWithNoWrap(ConstantRange(APInt(4, -5, true), APInt(4, 4)),
```
But this is... (also below).

File: llvm/unittests/IR/ConstantRangeTest.cpp [hunk 7]
```diff
@@ -1019,6 +1019,99 @@ TEST_F(ConstantRangeTest, Multiply) {
...
+          .multiplyWithNoWrap(ConstantRange(APInt(8, -1)), OBO::NoSignedWrap),
+      ConstantRange(APInt(8, 12), APInt(8, -2)));
+  EXPECT_EQ(ConstantRange(APInt(8, 254), APInt(8, 255))
+                .multiplyWithNoWrap(ConstantRange(APInt(8, 100), APInt(8, 121)),
+                                    OBO::NoSignedWrap),
+            ConstantRange(8, false));
+
+  auto CheckCorrectnessOnly = [](const ConstantRange &, const ConstantRange &) {
+    return false;
+  };
```
Comment (nikic) [inline llvm/unittests/IR/ConstantRangeTest.cpp:1071]: I'd move this to the top where the others are defined.

File: llvm/unittests/IR/ConstantRangeTest.cpp [hunk 8]
```diff
@@ -1019,6 +1019,99 @@ TEST_F(ConstantRangeTest, Multiply) {
...
+
+  TestBinaryOpExhaustive(
+      [](const ConstantRange &CR1, const ConstantRange &CR2) {
+        return CR1.multiplyWithNoWrap(CR2, OBO::NoUnsignedWrap);
+      },
+      [](const APInt &N1, const APInt &N2) -> std::optional<APInt> {
+        bool IsOverflow;
+        APInt Res = N1.umul_ov(N2, IsOverflow);
+        if (IsOverflow) {
```
Comment (nikic) [inline llvm/unittests/IR/ConstantRangeTest.cpp:1080]: No braces

**Comment by antoniofrighetto** [comment]
Argh, for some reason, I kept on considering them as if they were two different ranges :( Addressed reviews, thanks! (The `(4, true)` instance uses the CR constructor for empty/full set, not the APInt one)

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**